### DISCUSSION TITLE
How to Debug and Analyze Next.js Trace Files with Jaeger?

### DISCUSSION THREAD

**Discussion by smitpatel1010** [vercel/next.js#73412] [Help] [answered]
### Summary

I’m currently working with the Next.js trace file (`.next/trace`) generated during the build process. While this file provides valuable trace data, analyzing it in a structured tool like Jaeger would greatly enhance debugging and performance insights. However, I haven’t found any official or community-provided scripts/tools to convert the Next.js trace file into a Jaeger-compatible JSON format.

Here’s what I’m aiming for:
- A way to map the trace file's data (e.g., traceEvents) into Jaeger’s span model, including details like operation names, timestamps, durations, and parent-child relationships.
- A utility or guideline to bridge the format gap between Next.js traces and Jaeger’s trace ingestion requirements.

Questions:
1. Are there any existing tools, scripts, or practices for converting Next.js trace files to Jaeger-compatible formats?
2. Is there any documentation on the schema or structure of the Next.js trace file to help build such a conversion script?

**Comment by smitpatel1010** [comment]
@timneutkens I think you mentioned https://github.com/vercel/next.js/issues/66326#issuecomment-2177994701 here that you use some different tool. can you please mention which tool do you use?
`Upload the .next/trace file to https://gist.github.com/ -- Please don't run trace-to-tree yourself, as I use some other tools (e.g. Jaeger) that require the actual trace file.`

**Reply by m-sanders** [reply]
I am going down this rabbit hole as we speak. 

I am confused why Next.js has taken a "give a man a fish" approach here to debugging trace files rather than teaching us all to fish with some lovely documentation.

I've discovered a rust CLI tool in scripts that will send traces to Jaeger. I'm going to have a go at running Jaeger, compiling the script and attempt to ingest the trace file.

**Comment by m-sanders** [comment]
So this is actually relatively straight forward albeit undocumented. 

Prerequisites you've installed [`rust`](https://www.rust-lang.org/tools/install) and cloned the Next.js repo. I also have checked out the tag of the version I have installed just in case the format doesn't match.

1. Start the Jaeger all in one docker container:

https://www.jaegertracing.io/docs/2.4/getting-started/

```
docker run --rm --name jaeger \
  -p 16686:16686 \
  -p 4317:4317 \
  -p 4318:4318 \
  -p 5778:5778 \
  -p 9411:9411 \
  jaegertracing/jaeger:2.4.0
```

2. Remove you `.next` directory and run your app as outlined in the linked issue:

`NEXT_CPU_PROF=1 NEXT_TURBOPACK_TRACING=1 npm run dev`

3. Load a page your app. I used `curl http://localhost:3000/ > /dev/null`  to avoid making multiple requests on page load.

5. Ingest the trace file into Jaeger:

```
$ cd ~/code/next.js/scripts/send-trace-to-jaeger
$ cargo run ~/code/my-next-app/.next/trace
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.35s
     Running `/Users/me/code/next.js/target/debug/send-trace-to-jaeger trace`
Jaeger trace will be available on http://127.0.0.1:16686/trace/d98c390dcc1b3a5d
body = Ok("json: cannot unmarshal bool into Go struct field .Alias.tags of type string\n")
body = Ok("json: cannot unmarshal bool into Go struct field .Alias.tags of type string\n")
body = Ok("json: cannot unmarshal bool into Go struct field .Alias.tags of type string\n")
body = Ok("json: cannot unmarshal bool into Go struct field .Alias.tags of type string\n")
body = Ok("json: cannot unmarshal bool into Go struct field .Alias.tags of type string\n")
body = Ok("")
```

6. Open the link and view the trace in the web ui.

What I've discovered from this:

1. Our custom translations loader has become very expensive as we've added more translation strings and languages to our site. I will move to a generation step to convert these to json ahead of time rather than using a custom loader.
2. sass is 2-3x slower than css. I will expedite the migration from scss to css
3. That we load way too many modules. Time to move to Turbopack.

**Reply by timneutkens** [reply]
The Jaeger setup is undocumented intentionally. It's not the recommended way to interpret these. There are docs for the recommended way: https://nextjs.org/docs/app/guides/local-development#turbopack-tracing (also supports `.next/trace` for webpack traces)

**Comment by timneutkens** [comment, marked as answer]
Dev/Build tracing is now documented here: https://nextjs.org/docs/app/guides/local-development#turbopack-tracing

**Reply by m-sanders** [reply]
Thank you for the official solution!

I would still say that I've found the undocumented Jaeger approach to be easier to explore and get to the bottom of performance issues than the Turbopack tracing UI ;)

**Reply by 947776795** [reply]
<img width="1132" height="518" alt="image" src="https://github.com/user-attachments/assets/5e2af446-fc75-49ec-b9db-afa925b9788f" />
<img width="3220" height="254" alt="image" src="https://github.com/user-attachments/assets/761dd848-da97-4759-9027-f73c02207959" />
Why does it keep showing loading? How can I solve this problem?
